RecoHGCal/TICL/plugins/TracksterLinkingbySuperClustering.cc

```diff
@@ -0,0 +1,345 @@
+*/
+bool TracksterLinkingbySuperClustering::checkExplainedVarianceRatioCut(ticl::Trackster const& ts) const {
+  float explVar_denominator =
+      std::accumulate(std::begin(ts.eigenvalues()), std::end(ts.eigenvalues()), 0.f, std::plus<float>());
+  if (explVar_denominator != 0.) {
+    float explVarRatio = ts.eigenvalues()[0] / explVar_denominator;
+    if (ts.raw_energy() > explVarRatioCut_energyBoundary_)
+      return explVarRatio >= explVarRatioMinimum_highEnergy_;
+    else
+      return explVarRatio >= explVarRatioMinimum_lowEnergy_;
+  } else
+    return false;
+}
+
+/**
+ * resultTracksters : superclusters as tracksters (ie merging of tracksters that have been superclustered together)
+ * outputSuperclusters : same as linkedTracksterIdToInputTracksterId. Probably should use only one of the two.
+ * linkedTracksterIdToInputTracksterId : maps indices from resultTracksters back into input tracksters.
+ *    resultTracksters[i] has seed input.tracksters[linkedTracksterIdToInputTracksterId[i][0]], linked with tracksters input.tracksters[linkedTracksterIdToInputTracksterId[i][1..N]]
+*/
+void TracksterLinkingbySuperClustering::linkTracksters(
+    const Inputs& input,
+    std::vector<Trackster>& resultTracksters,
+    std::vector<std::vector<unsigned int>>& outputSuperclusters,
+    std::vector<std::vector<unsigned int>>& linkedTracksterIdToInputTracksterId) {
+  // For now we use all input tracksters for superclustering. At some point there might be a filter here for EM tracksters (electromagnetic identification with DNN ?)
+  auto const& inputTracksters = input.tracksters;
+  const unsigned int tracksterCount = inputTracksters.size();
+
+  std::unique_ptr<AbstractDNNInput> nnInput = makeDNNInputFromString(dnnVersion_);
+
+  /* Sorting tracksters by decreasing order of pT (out-of-place sort). 
+  inputTracksters[trackstersIndicesPt[0]], ..., inputTracksters[trackstersIndicesPt[N]] makes a list of tracksters sorted by decreasing pT
+  Indices into this pT sorted collection will have the suffix _pt. Thus inputTracksters[index] and inputTracksters[trackstersIndicesPt[index_pt]] are correct
+  */
+  std::vector<unsigned int> trackstersIndicesPt(inputTracksters.size());
+  std::iota(trackstersIndicesPt.begin(), trackstersIndicesPt.end(), 0);
+  std::stable_sort(
+      trackstersIndicesPt.begin(), trackstersIndicesPt.end(), [&inputTracksters](unsigned int i1, unsigned int i2) {
+        return inputTracksters[i1].raw_pt() > inputTracksters[i2].raw_pt();
+      });
+
+  /* Evaluate in minibatches since running with trackster count = 3000 leads to a short-lived ~15GB memory allocation
+  Also we do not know in advance how many superclustering candidate pairs there are going to be
+  So we pick (arbitrarily) a batch size of 10^5. It needs to be rounded to featureCount
+  */
+  const unsigned int miniBatchSize = static_cast<unsigned int>(1e5) / nnInput->featureCount() * nnInput->featureCount();
```

Comment:
would it be possible to choose the batch size at configuration?

Reply:
Done !